### PR TITLE
Add new set-generation cmd. Apiserver implementation to follow.

### PR DESCRIPTION
**Pull request by hmlanigan**
## Description of change

Initial implementations of the set-generation command. Apiserver side of the feature is not implemented.

## QA steps

1. export JUJU_DEV_FEATURE_FLAGS=generations
2. juju bootstrap; juju deploy ubuntu; juju deploy redis
2. juju set-generation 
unit and/or application names(s) must be specified <-- result
4. juju help set-generation
4. juju set-generation ubuntu/0
4. juju set-generation machine-0
ERROR invalid application or unit name "machine-0" <-- result
5.juju set-generation redis ubuntu/0

### CODE REVIEW

File: cmd/juju/model/setgeneration.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	"github.com/juju/gnuflag"
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/juju/api/modelgeneration"
+	jujucmd "github.com/juju/juju/cmd"
+	"github.com/juju/juju/cmd/modelcmd"
+)
+
+const (
+	setGenerationSummary = "Sets the generation for units and/or applications."
```
Comment (hmlanigan) [inline cmd/juju/model/setgeneration.go]: There is a trello card up to refine the help msgs.

**Comment by hmlanigan** [comment]
Changed set-generation to advance-generation

File: cmd/juju/model/advancegeneration.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+// SetFlags implements part of the cmd.Command interface.
+func (c *advanceGenerationCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.ModelCommandBase.SetFlags(f)
+}
+
+// Init implements part of the cmd.Command interface.
+func (c *advanceGenerationCommand) Init(args []string) error {
+	if len(args) == 0 {
+		return errors.Errorf("unit and/or application names(s) must be specified")
```
Comment (manadart) [inline cmd/juju/model/advancegeneration.go]: I assume the effect of advancing an application is that all units are advanced.

Perhaps down the line there should be some warning if an action at the CLI will auto-complete a generation. I see how advancing the application is useful - if an operator is satisfied with an initial canary unit(s) and they just want to move all others forward.

File: cmd/juju/model/advancegeneration_test.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+func (s *advanceGenerationSuite) runCommand(c *gc.C, api model.AdvanceGenerationCommandAPI, args ...string) (*cmd.Context, error) {
+	cmd := model.NewAdvanceGenerationCommandForTest(api, s.store)
+	return cmdtesting.RunCommand(c, cmd, args...)
+}
+
+func setUpAdvanceMocks(c *gc.C) (*gomock.Controller, *MockAdvanceGenerationCommandAPI) {
+	mockController := gomock.NewController(c)
+	mockAdvanceGenerationCommandAPI := NewMockAdvanceGenerationCommandAPI(mockController)
+	mockAdvanceGenerationCommandAPI.EXPECT().Close().Times(1)
```
Comment (manadart) [inline cmd/juju/model/advancegeneration_test.go]: Implicit.

File: cmd/juju/model/advancegeneration.go (hunk 2 above)
Comment (hmlanigan) [reply cmd/juju/model/advancegeneration.go]: Yes, advancing an application advances all the units.

**Comment by hmlanigan** [comment]
$$merge$$

**Pull request merged**